usr/opt/OSC200/usr/share/man/man3/ecvt.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: ecvt.3,v $ $Revision: 3.2.5.5 $ (DEC) $Date: 1994/01/13 16:24:31 $
.\"
.\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
.\" ALL RIGHTS RESERVED
.\" 
.\" (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 1.2
.\"
.TH ecvt 3
.SH NAME
.PP
\*Lecvt\*O, \*Lecvt_r\*O, \*Lfcvt\*O, \*Lfcvt_r\*O, \*Lgcvt\*O \- Convert a floating-point number to a string\&
.SH LIBRARY
.PP
\*OStandard C Library (\*Llibc.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <stdlib.h>\*O
.sp 0.4v
\*Lchar *ecvt(
.nL
	double \*Vvalue\*L,
.nL
	int \*Vnum_digits\*L,
.nL
	int *\*Vdecimal_ptr\*L,
.nL
	int *\*Vsign\*L);
.sp 0.4v
\*Lint ecvt_r(
.nL
	double \*Vvalue\*L,
.nL
	int \*Vnum_digits\*L,
.nL
	int *\*Vdecimal_ptr\*L,
.nL
	int *\*Vsign\*L,
.nL
	char *\*Vbuffer\*L,
.nL
	int \*Vlen \*L);
.sp 0.4v
char *fcvt(
.nL
	double \*Vvalue\*L,
.nL
	int \*Vnum_digits\*L,
.nL
	int *\*Vdecimal_ptr\*L,
.nL
	int *\*Vsign\*L);
.sp 0.4v
int fcvt_r(
.nL
	double \*Vvalue\*L,
.nL
	int \*Vnum_digits\*L,
.nL
	int *\*Vdecimal_ptr\*L,
.nL
	int *\*Vsign\*L,
.nL
	char *\*Vbuffer\*L,
.nL
	int \*Vlen \*L);
.sp 0.4v
char *gcvt(
.nL
	double \*Vvalue\*L,
.nL
	int \*Vnum_digits\*L,
.nL
	char *\*Vbuffer\*L);
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vvalue"
\*OSpecifies the double value to be converted.
.LI "\*Vnum_digits"
\*OSpecifies the number of digits in the resulting string.
.LI "\*Vdecimal_ptr"
\*OHolds the position of the decimal point
relative to the beginning of the string.
A negative number means
the decimal point is to the left of the digits given in the string.
.LI "\*Vsign"
\*OHolds a value of 0 (zero) if the value is positive or zero,
and a nonzero value
if it is negative.
.LI "\*Vbuffer"
\*OSpecifies the character array to be used as scratch space in calculations
and for storing the resulting string.
.LI "\*Vlen"
\*OSpecifies the length of \*Vbuffer\*O. Because \*Lecvt\*O temporarily stores
intermediate results of its calculations in the buffer, you must specify
a \*Vlen\*O greater than 370. A value of 400 is
recommended.
.LE
.SH DESCRIPTION
.PP
.iX "ecvt function"
.iX "ecvt_r function"
.iX "fcvt function"
.iX "fcvt_r function"
.iX "gcvt function"
.iX "floating-point number" "converting to a string"
The \*Lecvt(\|)\*O, \*Lfcvt(\|)\*O, and \*Lgcvt(\|)\*O
functions convert floating-point numbers to null-terminated strings.
.PP
The \*Lecvt(\|)\*O function converts the value specified by the \*Vvalue\*O
parameter
to a null-terminated string of length \*Vnum_digits\*O,
and returns a pointer to it.
The resulting low-order digit is rounded according to the
current rounding mode.
The \*Vdecimal_ptr\*O parameter is assigned to the position of the
decimal point relative to the position of the string.
The \*Vsign\*O parameter is assigned a value of 0 (zero) if
\*Vvalue\*O is positive or
zero, and a nonzero value if \*Vvalue\*O is negative.
The decimal point and sign are not included in the string.
.PP
The \*Lfcvt(\|)\*O function is the same as the \*Lecvt(\|)\*O function,
except that it rounds to the correct digit for outputting
\*Vnum_digits\*O of digits in C or FORTRAN F-format.  In the F-format,
\*Vnum_digits\*O is taken as the
number of digits desired after the decimal point.
.PP
The \*Lgcvt(\|)\*O function converts the value specified by the
\*Vvalue\*O parameter
to a null-terminated string, stores it in the array pointed to by
the \*Vbuffer\*O parameter, and then returns \*Vbuffer\*O.  The
\*Lgcvt(\|)\*O function
attempts to produce a string of \*Vnum_digits\*O significant
digits in FORTRAN F-format.  If this is not possible, then E-format
is used.  The
string is ready for printing, complete with minus sign, decimal point,
or exponent, as appropriate.  Trailing zeros are suppressed.
.PP
The \*Lecvt_r(\|)\*O and \*Lfcvt_r(\|)\*O functions are the reentrant
versions of the \*Lecvt(\|)\*O and \*Lfcvt(\|)\*O functions,
respectively. Upon successful completion, the resulting string is
stored in the array pointed to by the \*Vbuffer\*O parameter, and a
value of 0 (zero) is returned.
.SH NOTES
.PP
\*OIn the F-format, \*Vnum_digits\*O is the
number of digits desired after the decimal point.  Very large numbers
produce a very long string of digits before the decimal point, and
then \*Vnum_digits\*O of digits after the decimal point.  For large
numbers, it is preferable
to use the \*Lgcvt(\|)\*O or \*Lecvt(\|)\*O function so
that the E-format will be used.
.PP
The \*Lecvt(\|)\*O, \*Lfcvt(\|)\*O, and \*Lgcvt(\|)\*O
functions represent the following special values that are specified
in IEEE Standard 754-1985 for floating-point arithmetic according to
the following table.
.\" .TB "Representation of Special Floating-Point Values"
.\" previous line commented out until table re-coded 6/25/93
.TS
center tab(@);
lB  l B
l  l.
Value@Representation
_
Quiet NaN@NaNQ
Signalling NaN@NaNS
+Infinity@INF
-Infinity@-INF
.TE
.PP
The sign associated with each of these values is stored into the \*Vsign\*O
parameter.  Note, also, that in IEEE Floating Point, a value of
0 (zero) can be positive or negative, as set by the \*Vsign\*O parameter.
.SH CAUTIONS
.PP
The \*Lecvt(\|)\*O, \*Lfcvt(\|)\*O, and \*Lgcvt(\|)\*O functions
store the strings
in a static area of memory whose contents are overwritten each time
one of the functions is called.
.SH RETURN VALUES
.PP
The \*Lecvt(\|)\*O, \*Lfcvt(\|)\*O, and \*Lgcvt(\|)\*O
functions return the converted string.
.PP
Upon success, the \*Lecvt_r(\|)\*O and \*Lfcvt_r(\|)\*O functions
return 0 (zero).  Upon failure, they return -1 and \*Lerrno\*O may be set.
.SH ERRORS
.PP
If the \*Lecvt_r(\|)\*O or \*Lfcvt_r(\|)\*O function fails,
\*Lerrno\*O may be set to the following value:
.VL
.LI "[EINVAL]"
The \*Vbuffer\*O parameter is invalid or too small.
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Latof\*O(3),
\*Lprintf\*O(3),
\*Lscanf\*O(3).
.iX "ecvt function"
.iX "ecvt_r function"
.iX "fcvt function"
.iX "fcvt_r function"
.iX "gcvt function"
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/ecvt.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: ecvt.c,v $ $Revision: 4.2.9.3 $ (DEC) $Date: 1993/06/25 21:47:46 $";
#endif
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
/*
 */
/*
 * FUNCTIONS: ecvt, fcvt
 *
 * (C) COPYRIGHT International Business Machines Corp. 1988, 1989
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * ecvt.c	1.9  com/lib/c/cnv,3.1,8943 10/27/89 16:12:54
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#if defined(_THREAD_SAFE)
#pragma weak ecvt_r = __ecvt_r
#pragma weak fcvt_r = __fcvt_r
#endif
#if !defined(_THREAD_SAFE)
#pragma weak ecvt = __ecvt
#pragma weak fcvt = __fcvt
#endif
#endif
#include <math.h>
#include <fp.h>
#include <values.h>
#include <errno.h>
#include "print.h"

#include "ts_supp.h"

#define	NMAX	((DSIGNIF * 3 + 19)/10) /* restrict max precision */

#ifdef	_THREAD_SAFE
#define	RETURN(s)	{ strcpy(buf, s); return TS_SUCCESS;}
#else
#define	RETURN(s)	return s
static char	buf[NDIG];
#endif	/* _THREAD_SAFE */

static char *
cvt(double value, int mode, int ndigits, int *decpt, int *sign, char *buf)
{
	register char *p;
	register char *p_last;
	double valuem = value;		/* in memory for VALH to work */
	int f_flag;

	f_flag = (mode == 3) ? 1 : 0;
	p = &buf[0];
        p_last = &buf[ndigits];
        buf[0] = '\0';
        *decpt = 0;
        *sign = VALH(valuem) >> 31;

        if( IS_QNAN(valuem) )
                return("NaNQ");
        if( IS_NAN(valuem) )
                return("NaNS");
        if( IS_INF(valuem) )
                return("INF");
        if( IS_ZERO(valuem) ) {
                for ( ; p < p_last ; p++ )
                        *p = '0';
                *p = '\0';
                return(buf);
        }
        if (ndigits > 17)  {
                *decpt = _dtoa (buf, 17, value, f_flag) + 1;
        }
        else if (ndigits < 0) {
                *decpt = _dtoa (buf, 0, value, f_flag) + 1;
        }
        else {
                *decpt = _dtoa (buf, ndigits, value, f_flag) + 1;
        };
        *sign = buf[0] == '-';
        return(buf+1);
}


/*
 * NAME: ecvt
 *
 * FUNCTION: convert double to character string
 *
 * NOTES:
 *
 *	ecvt converts to decimal
 *	the number of digits is specified by ndigit
 *	decpt is set to the position of the decimal point
 *	sign is set to 0 for positive, 1 for negative
 *
 * RETURNS: a character string
 */
#ifdef _THREAD_SAFE
int
ecvt_r(double value, int ndigit, int *decpt, int *sign, char *buf, int len)
#else
char *
ecvt(double value, int ndigit, int *decpt, int *sign)
#endif	/* _THREAD_SAFE */
{
	TS_EINVAL((buf == 0) || (len < NDIG));
	RETURN(cvt(value, 2, ndigit, decpt, sign, buf));
}


/*
 * NAME: fcvt
 *
 * FUNCTION: convert double to character string if FORTRAN format
 *
 * NOTES:
 *
 * RETURNS: a character string
 *
 */
#ifdef _THREAD_SAFE
int
fcvt_r(double value, int ndigit, int *decpt, int *sign, char *buf, int len)

#else
char *
fcvt(double value, int ndigit, int *decpt, int *sign)
#endif	/* _THREAD_SAFE */
{
	TS_EINVAL((buf == 0) || (len < NDIG));
	RETURN( cvt(value, 3, ndigit, decpt, sign, buf));
}

